src/core/common/clearable.hpp

```diff
@@ -0,0 +1,60 @@
+/*
+ *  Copyright (c) 2020, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file includes definitions for Clearable class for OpenThread objects.
+ */
+
+#ifndef CLEARABLE_HPP_
+#define CLEARABLE_HPP_
+
+#include "openthread-core-config.h"
+
+#include <string.h>
+
+namespace ot {
+
+/**
+ * This template class defines a Clearable object which provides `Clear()` method.
+ *
+ * The `Clear` implementation simply sets all the bytes of a `Type` instance to zero.
+ *
+ * Users of this class should follow CRTP-style inheritance, i.e., the `Type` class itself should publicly inherit
+ * from `Clearable<Type>`.
+ *
+ */
+template <typename Type> class Clearable
+{
+public:
+    void Clear(void) { memset(reinterpret_cast<void *>(this), 0, sizeof(Type)); }
```

Comment:
```suggestion
    void Clear(void) { *this = Type(); }
```
Safer than `memset`.

Reply:
If Type() calls `Clear` and `Clear` calls Type(), would that be a infinite loop?

Reply:
`Type()` shouldn't call `Clear`. Is there any case that this is required?

Reply:
see https://github.com/openthread/openthread/blob/8e69b5acb7003223f517dff2bc9ef720ac4c9ff8/src/core/net/socket.hpp#L215

Reply:
In this case the constructor can be changed to `SockAddr(void) : otSockAddr() {}` without changing any behaviors.
But these changes are not trivial. If it's impractical to change all such calls, I'm OK with the `memset` solution.

Reply:
This PR is only refactoring existing code by adding a common sub-class to avoid repeating the same logic in many different classes. Existing code uses `memset` already.

Here are some reasons why I would say `memset` is preferred:
- I think `Type()` does ensure zero initialization since C++11 (https://en.cppreference.com/w/cpp/language/zero_initialization). Note that OT still only requires C++03.
- It adds a restriction for the `Type` to have a public default constructor `Type()` so that it can inherit from `Clearable` (unnecessarily additional requirement) and that the default constructor should set all fields to zero. 
- OT is network stack and therefore is being built with a lot of older tool-chains/compliers. An older complier (not performing good optimization) may treat `*this = Type()` to create a new instance and zero initialize it and then copy it to `*this` (which adds an extra copying). `memset` approach would not rely on compiler optimization.